src/gpuarray_util.c

```diff
@@ -130,3 +130,45 @@ int gpuarray_type_flags(int init, ...) {
   va_end(ap);
   return flags;
 }
+
+static inline void shiftdown(ssize_t *base, unsigned int i, unsigned int nd) {
+  if (base != NULL)
+    memmove(&base[i], &base[i+1], (nd - i - 1)*sizeof(size_t));
+}
+
+void gpuarray_elemwise_collapse(unsigned int n, unsigned int *_nd,
```

Comment:
In the python version, there is a parameter broadcast. When True, it set the strides to 0 for dimensions that need to be broadcasted. Can you confirm this is already done elsewhere and don't need to be replicated here?

Otherwise, this PR seem good. Can you make the PR on Theano that use it? This would allow to test it more before merging.